Public/Src/Engine/Processes/SharedOpaqueOutputLogger.cs

```diff
@@ -0,0 +1,233 @@
+        {
+            Contract.Requires(searchRootDirectory.IsValid);
+
+            var semiStableHashX16 = string.Format(CultureInfo.InvariantCulture, "{0:X16}", process.SemiStableHash);
+            var subDirName = semiStableHashX16.Substring(0, 3);
+
+            return searchRootDirectory.Combine(pathTable, subDirName).Combine(pathTable, $"{LogFilePrefix}{semiStableHashX16}{LogFileSuffix}").ToString(pathTable);
+        }
+
+        /// <summary>
+        /// Finds and returns all output log files that exist in directory denoted by <paramref name="directory"/>
+        /// </summary>
+        /// <remarks>
+        /// CODESYNC: must be consistent with <see cref="GetOutputLogFileForProcess(PathTable, AbsolutePath, Process)"/>
+        /// </remarks>
+        public static string[] FindAllProcessPipOutputLogFiles(string directory)
+        {
+            return Directory.Exists(directory)
+                ? Directory.EnumerateFiles(directory, $"{LogFilePrefix}*{LogFileSuffix}", SearchOption.AllDirectories).ToArray()
+                : CollectionUtilities.EmptyArray<string>();
+        }
+
+        /// <summary>
+        /// Returns all paths recorded in the <paramref name="filePath"/> file.
+        /// </summary>
+        /// <remarks>
+        /// Those paths are expected to be absolute paths of files/directories that were written to by the previous build.
+        ///
+        /// NOTE: this method does not validate the recorded paths in any way.  That means that each returned string may be
+        ///   - a path pointing to an absent file
+        ///   - a path pointing to a file
+        ///   - a path pointing to a directory.
+        /// 
+        /// NOTE: unless the log file was corrupted (i.e., it was produced by an instance of this class):
+        ///   - the strings in the returned enumerable are all legal paths
+        ///   - the returned collection does not contain any duplicates
+        /// </remarks>
+        public static IEnumerable<string> ReadRecordedPathsFromSharedOpaqueOutputLog(string filePath)
+        {
+            using (var fileStream = new FileStream(filePath, FileMode.Open, FileAccess.Read, FileShare.Read | FileShare.Delete))
+            using (var bxlReader = new BuildXLReader(stream: fileStream, debug: false, leaveOpen: false))
+            {
+                string nextString = null;
+                while ((nextString = ReadStringOrNull(bxlReader)) != null)
+                {
+                    yield return nextString;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Same as <see cref="ReadRecordedPathsFromSharedOpaqueOutputLog(string)"/> except that all exceptions are wrapped in <see cref="BuildXLException"/>
+        /// </summary>
+        public static string[] ReadRecordedPathsFromSharedOpaqueOutputLogWrapExceptions(string filePath)

```

Comment:
Why do we need two public user facing API? Why can't we just have `ReadRecordedPathsFromSharedOpaqueOutputLogWrapExceptions` ?

Reply:
will change

Reply:
I removed the WrapExceptions method.

The reason why left the other one is because the other one can be a true iterator (i.e., it's not possible to yield a value from a try/catch block)